src/features/colinks/BuyButton.tsx

```diff
@@ -64,30 +65,50 @@ export const BuyButton = ({
   };
 
   const buyLinkWithContract = async (
-    signedContract: CoLinks,
+    coLinksWithWallet: CoLinksWithWallet,
     chainId: string
   ) => {
     try {
-      console.log({ signedContract, chainId });
-      assert(chainId);
+      assert(chainId, 'BuyButton chainId is required');
+      assert(currentUserAddress, 'address is required');
       setAwaitingWallet(true);
 
-      const value = await signedContract.getBuyPriceAfterFee(target, 1);
+      const value = await coLinksWithWallet.read.getBuyPriceAfterFee([
+        target as Address,
+        BigInt(1),
+      ] as const);
 
-      console.log('try buy');
+      console.log('buyLinkWithContract', { value });
+
+      const { request } = await coLinksWithWallet.simulate.buyLinks(
+        [target as Address, BigInt(1)] as const,
+        {
+          account: currentUserAddress,
+          value,
+          chain: wagmiChain,
+        }
+      );
+
+      console.log('Simulation successful. Estimated gas:', { request });
+
+      console.log('try buy', { value, coLinksWithWallet, currentUserAddress });
       const { receipt, error /*, tx*/ } = await sendAndTrackTx(
         () => {
-          console.log({ signedContract });
-          return signedContract.buyLinks(target, 1, {
-            value,
-          });
+          return coLinksWithWallet.write.buyLinks(
+            [target as Address, 1n] as const,
+            {
+              value,
+              account: currentUserAddress,
+              chain: wagmiChain,
```

Comment:
curious that the chain has to be passed in, seems like we should do this when we build CoLinksWithWallet?

Reply:
maybe make custom wrapper to encapsulate logic

Reply:
follow up task